File: tests/test_gate_api.js
/**
 * Gate API End-to-End Test
 *
 * Tests the full gate API flow:
 * 1. Create an API key for a project
 * 2. Evaluate test cases through the gate
 * 3. Verify responses have correct structure
 * 4. Test error cases (invalid key, missing fields)
 * 5. Test rate limiting
 *
 * Usage:
 *   node sdk/tests/test_gate_api.js [baseUrl] [projectId]
 *
 * Prerequisites:
 *   - Server running at baseUrl (default: http://localhost:8080)
 *   - A project with taxonomy template and review_mode='automated' exists
 *   - A logged-in session cookie (for key creation) OR pass --skip-create to use existing key
 */

const testData = require('./gate_test_data.json');

const BASE_URL = process.argv[2] || 'http://localhost:8080';
const PROJECT_ID = parseInt(process.argv[3]) || 1;

let API_KEY = null;
let RESULTS = { passed: 0, failed: 0, skipped: 0, errors: [] };

function log(icon, msg) { console.log(`${icon} ${msg}`); }
function pass(name) { RESULTS.passed++; log('✅', name); }
function fail(name, reason) { RESULTS.failed++; RESULTS.errors.push({ name, reason }); log('❌', `${name}: ${reason}`); }
function skip(name, reason) { RESULTS.skipped++; log('⏭️', `${name}: ${reason}`); }

async function request(method, path, body, headers = {}) {
    const url = `${BASE_URL}${path}`;
    const opts = {
        method,
        headers: { 'Content-Type': 'application/json', ...headers },
    };
    if (body) opts.body = JSON.stringify(body);

    const res = await fetch(url, opts);
    const data = await res.json().catch(() => ({ error: `HTTP ${res.status}` }));
    return { status: res.status, ...data };
}

// ===== Test: Key Management =====

async function testCreateKey() {
    log('🔑', 'Testing key creation...');

    // This requires session auth — we'll create a key via direct DB insert for testing
    const crypto = require('crypto');
    const fullKey = `bk_gate_${crypto.randomBytes(20).toString('hex')}`;
    const keyHash = crypto.createHash('sha256').update(fullKey).digest('hex');
    const keyPrefix = fullKey.substring(0, 16);

    // Insert directly via the gate context
    try {
        process.env.master = 'development';
        const gateContext = require('../../components/gate/app/models/gateContext');
        const ctx = new gateContext();

        // Wait for connection
        await new Promise(resolve => setTimeout(resolve, 2000));

        await ctx._execute(
            "INSERT INTO `GateApiKey` (`project_id`, `user_id`, `name`, `key_hash`, `key_prefix`, `allowed_origins`, `rate_limit_max`, `rate_limit_window_ms`, `is_active`, `last_used_at`, `usage_count`, `created_at`, `updated_at`) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
            [PROJECT_ID, 1, 'Test Key', keyHash, keyPrefix, '[]', 100, 60000, 1, null, 0, Date.now().toString(), Date.now().toString()]
        );

        API_KEY = fullKey;
        pass('Key created via direct insert');
        log('🔑', `Key: ${keyPrefix}...`);
    } catch (e) {
        fail('Key creation', e.message);
    }
}

// ===== Test: Evaluate endpoint =====

async function testEvaluateNoKey() {
    const res = await request('POST', '/api/v1/gate/evaluate', { input: 'test', output: 'test' });
    if (res.error && (res.status === 401 || res.error.includes('API key'))) {
        pass('Evaluate without key returns 401');
    } else {
        fail('Evaluate without key', `Expected 401, got: ${JSON.stringify(res).substring(0, 200)}`);
    }
}

async function testEvaluateInvalidKey() {
    const res = await request('POST', '/api/v1/gate/evaluate',
        { input: 'test', output: 'test' },
        { 'X-API-Key': 'bk_gate_invalid_key_12345678901234567890' }
    );
    if (res.error && res.error.includes('Invalid')) {
        pass('Evaluate with invalid key returns error');
    } else {
        fail('Evaluate with invalid key', `Expected error, got: ${JSON.stringify(res).substring(0, 200)}`);
    }
}

async function testEvaluateMissingBody() {
    if (!API_KEY) { skip('Evaluate missing body', 'No API key'); return; }
    const res = await request('POST', '/api/v1/gate/evaluate', {}, { 'X-API-Key': API_KEY });
    if (res.error) {
        pass('Evaluate with empty body returns error');
    } else {
        fail('Evaluate with empty body', `Expected error, got: ${JSON.stringify(res).substring(0, 200)}`);
    }
}

async function testEvaluateWithData() {
    if (!API_KEY) { skip('Evaluate with data', 'No API key'); return; }

    const testCase = testData.test_cases[0]; // Good customer support response
    log('📝', `Testing: ${testCase.name}`);

    const res = await request('POST', '/api/v1/gate/evaluate', {
        input: testCase.input,
        output: testCase.output,
        context: testCase.context,
        metadata: testCase.metadata
    }, { 'X-API-Key': API_KEY });

    if (!res.success && res.error) {
        // If it fails due to no taxonomy/model, that's a setup issue, not a code bug
        if (res.error.includes('taxonomy') || res.error.includes('model') || res.error.includes('not active')) {
            skip(`Evaluate: ${testCase.name}`, `Project setup needed: ${res.error}`);
            return;
        }
        fail(`Evaluate: ${testCase.name}`, res.error);
        return;
    }

    // Validate response structure
    const requiredFields = ['decision', 'risk', 'flags', 'policy_action', 'audit_id', 'task_id'];
    const missingFields = requiredFields.filter(f => res[f] === undefined);

    if (missingFields.length > 0) {
        fail(`Response structure: ${testCase.name}`, `Missing fields: ${missingFields.join(', ')}`);
    } else {
        pass(`Response structure: ${testCase.name}`);
    }

    // Validate decision values
    const validDecisions = ['allow', 'flag', 'block', 'queued'];
    if (!validDecisions.includes(res.decision)) {
        fail(`Decision value: ${testCase.name}`, `Invalid decision: ${res.decision}`);
    } else {
        pass(`Decision value: ${testCase.name} → ${res.decision}`);
    }

    // Validate risk values
    const validRisks = ['low', 'medium', 'high', null];
    if (!validRisks.includes(res.risk)) {
        fail(`Risk value: ${testCase.name}`, `Invalid risk: ${res.risk}`);
    } else {
        pass(`Risk value: ${testCase.name} → ${res.risk}`);
    }

    // Validate policy_action
    const validActions = ['allow', 'review', 'block', 'require_sme', 'allow_with_warning', 'queued'];
    if (!validActions.includes(res.policy_action)) {
        fail(`Policy action: ${testCase.name}`, `Invalid policy_action: ${res.policy_action}`);
    } else {
        pass(`Policy action: ${testCase.name} → ${res.policy_action}`);
    }

    // Log full response for inspection
    log('📊', `Full response:`);
    console.log(JSON.stringify(res, null, 2));
}

async function testEvaluateAllCases() {
    if (!API_KEY) { skip('Evaluate all cases', 'No API key'); return; }

    log('', '\n========== Running all test cases ==========\n');

    for (const testCase of testData.test_cases) {
        log('📝', `Testing: ${testCase.name}`);

        try {
            const res = await request('POST', '/api/v1/gate/evaluate', {
                input: testCase.input,
                output: testCase.output,
                context: testCase.context || {},
                metadata: testCase.metadata || {}
            }, { 'X-API-Key': API_KEY });

            if (!res.success) {
                if (res.error?.includes('taxonomy') || res.error?.includes('model') || res.error?.includes('not active')) {
                    skip(testCase.name, `Project setup: ${res.error}`);
                    continue;
                }
                fail(testCase.name, res.error || 'Unknown error');
                continue;
            }

            // Log result summary
            const icon = res.decision === 'allow' ? '🟢' : res.decision === 'block' ? '🔴' : '🟡';
            log(icon, `  Decision: ${res.decision} | Risk: ${res.risk} | Action: ${res.policy_action} | Confidence: ${res.confidence} | Flags: ${(res.flags || []).join(', ') || 'none'}`);
            if (res.rationale) log('💬', `  Rationale: ${res.rationale.substring(0, 120)}...`);
            log('⏱️', `  Evaluation: ${res.evaluation_ms}ms | Audit: ${res.audit_id}`);

            pass(`${testCase.name} → ${res.decision} (expected: ${testCase.expected_decision})`);
        } catch (e) {
            fail(testCase.name, e.message);
        }
    }
}

// ===== Test: SDK =====

async function testNodeSdk() {
    if (!API_KEY) { skip('Node SDK', 'No API key'); return; }

    log('📦', 'Testing Node.js SDK...');

    try {
        const { BookbagClient } = require('../node/src');
        const client = new BookbagClient({ apiKey: API_KEY, baseUrl: BASE_URL });

        const result = await client.gate.evaluate({
            input: 'What is my account balance?',
            output: 'Your balance is $5,230.00.',
            context: { channel: 'support_chat' },
            metadata: { session_id: 'sdk_test' }
        });

        if (result.decision && result.audit_id && result.policy_action) {
            pass(`Node SDK evaluate → ${result.decision}`);
        } else {
            fail('Node SDK evaluate', `Missing fields in response: ${JSON.stringify(result).substring(0, 200)}`);
        }
    } catch (e) {
        if (e.message?.includes('taxonomy') || e.message?.includes('model')) {
            skip('Node SDK', `Project setup: ${e.message}`);
        } else {
            fail('Node SDK', e.message);
        }
    }
}

// ===== Test: Queue Manager =====

async function testQueueManager() {
    log('🔄', 'Testing GateQueueManager...');

    const GateQueueManager = require('../../components/gate/app/service/gateQueueManager');
    const qm = new GateQueueManager({ maxGlobal: 3, maxPerProject: 2 });

    // Test basic execution
    const result = await qm.execute(1, async () => 'hello');
    if (result === 'hello') {
        pass('QueueManager basic execution');
    } else {
        fail('QueueManager basic execution', `Expected 'hello', got '${result}'`);
    }

    // Test concurrency limiting
    let concurrent = 0;
    let maxConcurrent = 0;
    const tasks = Array.from({ length: 6 }, (_, i) =>
        qm.execute(1, async () => {
            concurrent++;
            maxConcurrent = Math.max(maxConcurrent, concurrent);
            await new Promise(r => setTimeout(r, 50));
            concurrent--;
            return i;
        })
    );

    const results = await Promise.all(tasks);
    if (maxConcurrent <= 2) {
        pass(`QueueManager per-project limit (max concurrent: ${maxConcurrent}, limit: 2)`);
    } else {
        fail(`QueueManager per-project limit`, `Max concurrent ${maxConcurrent} exceeded limit 2`);
    }

    if (results.length === 6) {
        pass('QueueManager all tasks completed');
    } else {
        fail('QueueManager completion', `Expected 6 results, got ${results.length}`);
    }

    // Test metrics
    const metrics = qm.getMetrics();
    if (metrics.totalProcessed > 0) {
        pass(`QueueManager metrics: ${metrics.totalProcessed} processed, ${metrics.totalQueued} queued`);
    } else {
        fail('QueueManager metrics', 'No processing recorded');
    }
}

// ===== Test: Policy Service =====

async function testPolicyService() {
    log('📋', 'Testing PolicyService...');

    const PolicyService = require('../../components/qa/app/service/policyService');
    const ps = new PolicyService(null);

    // Test default policy
    const defaults = ps.getDefaultPolicy();
    if (defaults.routing && defaults.risk_rules && defaults.export_rules) {
        pass('PolicyService default policy');
    } else {
        fail('PolicyService default policy', 'Missing fields');
    }

    // Test validation
    const { valid, errors } = ps.validatePolicy(defaults);
    if (valid) {
        pass('PolicyService validate default policy');
    } else {
        fail('PolicyService validate', errors.join(', '));
    }

    // Test gate policy - safe_to_deploy
    const project = {
        policy_config: {
            risk_rules: [{ risk: 'high', action: 'require_sme' }],
            flag_rules: [{ flag: 'hallucination', action: 'require_qa' }],
            auto_decision_rules: [],
            enforcement_mode: 'advisory'
        },
        gate_config: {}
    };

    const safeEval = {
        production_verdict: 'safe_to_deploy',
        failure_categories: [],
        confidence: 0.95
    };
    const safeResult = ps.applyGatePolicy(project, safeEval);
    if (safeResult.decision === 'allow' && safeResult.risk === 'low' && safeResult.policy_action === 'allow') {
        pass('PolicyService safe_to_deploy → allow/low/allow');
    } else {
        fail('PolicyService safe_to_deploy', JSON.stringify(safeResult));
    }

    // Test gate policy - blocked with high risk
    const blockedEval = {
        production_verdict: 'blocked',
        failure_categories: ['hallucination', 'pii_exposure'],
        confidence: 0.3
    };
    const blockedResult = ps.applyGatePolicy(project, blockedEval);
    if (blockedResult.decision === 'block' && blockedResult.risk === 'high') {
        pass(`PolicyService blocked → block/high/${blockedResult.policy_action}`);
    } else {
        fail('PolicyService blocked', JSON.stringify(blockedResult));
    }

    // Test gate policy - flag with hallucination
    const flagEval = {
        production_verdict: 'needs_fix',
        failure_categories: ['hallucination'],
        confidence: 0.6
    };
    const flagResult = ps.applyGatePolicy(project, flagEval);
    if (flagResult.decision === 'flag' && flagResult.matched_rules.length > 0) {
        pass(`PolicyService needs_fix+hallucination → flag/${flagResult.risk}/${flagResult.policy_action}`);
    } else {
        fail('PolicyService flag+hallucination', JSON.stringify(flagResult));
    }

    // Test export policy
    const exportResult = ps.applyExportPolicy(project, { production_verdict: 'blocked', failure_categories: '["pii_exposure"]' });
    if (exportResult.export_set === 'safety') {
        pass('PolicyService export blocked → safety set');
    } else {
        fail('PolicyService export', JSON.stringify(exportResult));
    }
}

// ===== Main =====

async function main() {
    console.log('\n🚀 Bookbag Gate API Test Suite');
    console.log(`   Base URL: ${BASE_URL}`);
    console.log(`   Project ID: ${PROJECT_ID}`);
    console.log('');

    // Unit tests (no server needed)
    await testQueueManager();
    console.log('');
    await testPolicyService();
    console.log('');

    // Integration tests (need server + DB)
    await testCreateKey();
    console.log('');
    await testEvaluateNoKey();
    await testEvaluateInvalidKey();
    await testEvaluateMissingBody();
    console.log('');

    // Full evaluation tests (need server + project with taxonomy + LLM)
    await testEvaluateWithData();
    console.log('');
    await testEvaluateAllCases();
    console.log('');

    // SDK tests
    await testNodeSdk();
    console.log('');

    // Summary
    console.log('========== Results ==========');
    console.log(`✅ Passed: ${RESULTS.passed}`);
    console.log(`❌ Failed: ${RESULTS.failed}`);
    console.log(`⏭️ Skipped: ${RESULTS.skipped}`);

    if (RESULTS.errors.length > 0) {
        console.log('\nFailures:');
        RESULTS.errors.forEach(e => console.log(`  - ${e.name}: ${e.reason}`));
    }

    console.log('');
    process.exit(RESULTS.failed > 0 ? 1 : 0);
}

main().catch(e => { console.error('Test runner error:', e); process.exit(1); });

File: tests/test_gate_api.py
#!/usr/bin/env python3
"""
Gate API Test Suite (Python SDK)

Tests the Gate API using the Python SDK.

Usage:
    python sdk/tests/test_gate_api.py [api_key] [base_url]

Prerequisites:
    - Server running at base_url
    - A valid gate API key for an automated project
"""

import sys
import os
import json

# Add SDK to path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'python'))

from bookbag import BookbagClient
from bookbag.exceptions import BookbagError, AuthenticationError, RateLimitError

API_KEY = sys.argv[1] if len(sys.argv) > 1 else 'bk_gate_YOUR_KEY_HERE'
BASE_URL = sys.argv[2] if len(sys.argv) > 2 else 'http://localhost:8080'

passed = 0
failed = 0
skipped = 0

def log(icon, msg):
    print(f"{icon} {msg}")

def test_pass(name):
    global passed
    passed += 1
    log("✅", name)

def test_fail(name, reason):
    global failed
    failed += 1
    log("❌", f"{name}: {reason}")

def test_skip(name, reason):
    global skipped
    skipped += 1
    log("⏭️", f"{name}: {reason}")


def test_invalid_key():
    """Test that invalid API key raises AuthenticationError."""
    try:
        client = BookbagClient(api_key="bk_gate_invalid", base_url=BASE_URL)
        client.gate.evaluate(input="test", output="test")
        test_fail("Invalid key", "Should have raised AuthenticationError")
    except AuthenticationError:
        test_pass("Invalid key raises AuthenticationError")
    except Exception as e:
        # Connection error means server isn't running
        if "Connection" in str(e) or "urlopen" in str(e):
            test_skip("Invalid key", f"Server not reachable: {e}")
        else:
            test_fail("Invalid key", str(e))


def test_missing_key():
    """Test that missing API key raises error."""
    try:
        BookbagClient(api_key="")
        test_fail("Missing key", "Should have raised error")
    except AuthenticationError:
        test_pass("Missing key raises AuthenticationError")


def test_evaluate():
    """Test evaluation with valid key."""
    if API_KEY == 'bk_gate_YOUR_KEY_HERE':
        test_skip("Evaluate", "No API key provided")
        return

    client = BookbagClient(api_key=API_KEY, base_url=BASE_URL)

    # Load test data
    test_data_path = os.path.join(os.path.dirname(__file__), 'gate_test_data.json')
    with open(test_data_path) as f:
        test_cases = json.load(f)['test_cases']

    for tc in test_cases[:3]:  # Test first 3 cases
        try:
            result = client.gate.evaluate(
                input=tc['input'],
                output=tc['output'],
                context=tc.get('context', {}),
                metadata=tc.get('metadata', {})
            )

            # Validate response structure
            assert result.decision in ('allow', 'flag', 'block', 'queued'), f"Bad decision: {result.decision}"
            assert result.audit_id, "Missing audit_id"
            assert result.policy_action in ('allow', 'review', 'block', 'require_sme', 'allow_with_warning', 'queued'), f"Bad policy_action: {result.policy_action}"

            icon = '🟢' if result.decision == 'allow' else '🔴' if result.decision == 'block' else '🟡'
            log(icon, f"  {tc['name']}: decision={result.decision} risk={result.risk} action={result.policy_action} confidence={result.confidence}")
            test_pass(f"Evaluate: {tc['name']}")

        except BookbagError as e:
            if 'taxonomy' in str(e).lower() or 'model' in str(e).lower():
                test_skip(f"Evaluate: {tc['name']}", f"Project setup: {e}")
            else:
                test_fail(f"Evaluate: {tc['name']}", str(e))
        except Exception as e:
            if "Connection" in str(e) or "urlopen" in str(e):
                test_skip(f"Evaluate: {tc['name']}", "Server not reachable")
                break
            else:
                test_fail(f"Evaluate: {tc['name']}", str(e))


def test_sdk_result_object():
    """Test GateResult dataclass."""
    from bookbag.gate import GateResult

    result = GateResult(
        decision="allow",
        risk="low",
        flags=[],
        policy_action="allow",
        enforced=False,
        audit_id="test_123",
        task_id=1,
        confidence=0.95,
        scores={"correctness": 5},
        rationale="Looks good",
        evaluation_ms=1234
    )

    assert result.decision == "allow"
    assert result.risk == "low"
    assert result.confidence == 0.95
    assert result.audit_id == "test_123"
    assert result.evaluation_ms == 1234
    test_pass("GateResult dataclass")


def main():
    print(f"\n🐍 Bookbag Gate API Python SDK Tests")
    print(f"   Base URL: {BASE_URL}")
    print(f"   API Key: {API_KEY[:16]}..." if len(API_KEY) > 16 else f"   API Key: {API_KEY}")
    print()

    # Unit tests (no server needed)
    test_missing_key()
    test_sdk_result_object()
    print()

    # Integration tests (need server)
    test_invalid_key()
    print()

    # Evaluation tests (need server + project + LLM)
    test_evaluate()
    print()

    # Summary
    print("========== Results ==========")
    print(f"✅ Passed: {passed}")
    print(f"❌ Failed: {failed}")
    print(f"⏭️ Skipped: {skipped}")

    sys.exit(1 if failed > 0 else 0)


if __name__ == "__main__":
    main()
